realm/realm-library/src/main/java/io/realm/RealmMapEntrySet.java

```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Override
+    public boolean retainAll(@NotNull Collection<?> c) {
+        throw new UnsupportedOperationException("This set is immutable and cannot be modified.");
+    }
+
+    @Override
+    public boolean removeAll(@NotNull Collection<?> c) {
+        throw new UnsupportedOperationException("This set is immutable and cannot be modified.");
+    }
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException("This set is immutable and cannot be modified.");
+    }
+
+    private static <K, V> Iterator<Map.Entry<K, V>> iteratorFactory(IteratorType iteratorType,
+                                                                    OsMap osMap,
+                                                                    BaseRealm baseRealm,
+                                                                    @Nullable ClassContainer classContainer) {
+        switch (iteratorType) {
+            case PRIMITIVE:
+                return new PrimitiveValueIterator<>(osMap, baseRealm);
+            case MIXED:
+                //noinspection unchecked
+                return (EntrySetIterator<K, V>) new MixedValueIterator<K>(osMap, baseRealm);
+            case OBJECT:
+                if (classContainer == null) {
+                    throw new IllegalArgumentException("Missing class container when creating RealmModelValueIterator.");
+                }
+                return new RealmMapEntrySet.RealmModelValueIterator<>(osMap, baseRealm, classContainer);
+            default:
+                throw new IllegalArgumentException("Invalid iterator type.");
+        }
+    }
+
+    private abstract static class EntrySetIterator<K, V> implements Iterator<Map.Entry<K, V>> {
+
+        protected final OsMap osMap;
+        protected final BaseRealm baseRealm;
+
+        private int pos = -1;
+
+        public EntrySetIterator(OsMap osMap, BaseRealm baseRealm) {
+            this.osMap = osMap;
+            this.baseRealm = baseRealm;
+        }
+
+        protected abstract Map.Entry<K, V> getEntryInternal(int position);
+
+        @Override
+        public boolean hasNext() {
+            return pos + 1 < osMap.size();
+        }
+
+        @Override
+        public Map.Entry<K, V> next() {
+            pos++;
+            long size = osMap.size();
+            if (pos >= size) {
+                throw new NoSuchElementException("Cannot access index " + pos + " when size is " + size +
+                        ". Remember to check hasNext() before using next().");
+            }
+
+            return getEntryInternal(pos);
+        }
+    }
+
+    private static class PrimitiveValueIterator<K, V> extends EntrySetIterator<K, V> {
+
+        public PrimitiveValueIterator(OsMap osMap, BaseRealm baseRealm) {
+            super(osMap, baseRealm);
+        }
+
+        @Override
+        protected Map.Entry<K, V> getEntryInternal(int position) {
```

Comment:
Here we are leveraging Mixed to access primitive types, which has some performance penalty as we have to cross the JNI 2 extra times. Did we think about implementing direct primitive access?

Reply:
True that. I had left it like this as it worked fine for `copyToRealm` from the proxy, but I thought of it with the intention of improving it later. I have created an internal task for it: RJAVA-886

Reply:
I ended up fixing it here.